File: 0046-permutations/0046-permutations.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permute = function(nums) {

    const results = [];

    const compute = (current) => {
        if (current == nums.length) return results.push([...nums]);

        for (let pointer = current; pointer < nums.length;pointer++) {
            const temp = nums[current];
            nums[current] = nums[pointer];
            nums[pointer] = temp;
            compute(current+1);
            nums[pointer] = nums[current];
            nums[current] = temp;
        }     
    }
    compute(0);
    return results;
    
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        results = []
        def solver(used,remaining):
            if len(remaining) == 0: results.append(used.copy())
            for i in range(0,len(remaining)):
                r = remaining[i]
                used.append(r)
                remaining.pop(i)
                solver(used,remaining)
                used.pop()
                remaining.insert(i,r)
        
        solver([],nums)
        return results
